### ISSUE TITLE
`torch.unique` produces different from `numpy` results with inputs of `dim >=3` and specified `dim`.

### ISSUE THREAD

**Issue by Verma-Rajat** [pytorch/pytorch#101681]
### 🐛 Describe the bug

torch.unique documentation states
"Currently in the CUDA implementation and the CPU implementation when dim is specified, torch.unique always sort the tensor at the beginning regardless of the sort argument."

However for below example, it can be seen, output is not sorted correctly.

```python
import torch
x = torch.tensor([[[[1, 0, 1, 0, 1, 1],
         [0, 1, 1, 0, 1, 1]],

        [[0, 1, 1, 0, 0, 1],
         [0, 0, 0, 1, 0, 0]]],


       [[[0, 1, 0, 1, 1, 1],
         [0, 1, 1, 0, 1, 1]],

        [[0, 0, 1, 1, 0, 1],
         [1, 1, 0, 0, 0, 0]]]])
y = torch.unique(x,dim=-1)
print(y)
```

Output I get is:
0, 0, 1, **1**, **1**, 1
0, 1, 0, **1**, **1**, 1
0, 1, 0, **1**, **0**, 1
1, 0, 0, 0, 0, 0
1, 1, 0, 0, 1, 1
0, 1, 0, 1, 1, 1
1, 0, 0, 1, 0, 1
0, 1, 1, 0, 0, 0

It can be seen, that 3rd and 4th columns are not in correct order. 4th column (containing 1,1,0...) should have came before 3rd one (containing 1,1,1...)

### Versions

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:               None
        - available:         False
        - version:           11.7
* Lightning:
        - torch:             2.0.1
* Packages:
        - cmake:             3.26.3
        - filelock:          3.12.0
        - jinja2:            3.1.2
        - lit:               16.0.3
        - markupsafe:        2.1.2
        - mpmath:            1.3.0
        - networkx:          3.1
        - numpy:             1.24.3
        - nvidia-cublas-cu11: 11.10.3.66
        - nvidia-cuda-cupti-cu11: 11.7.101
        - nvidia-cuda-nvrtc-cu11: 11.7.99
        - nvidia-cuda-runtime-cu11: 11.7.99
        - nvidia-cudnn-cu11: 8.5.0.96
        - nvidia-cufft-cu11: 10.9.0.58
        - nvidia-curand-cu11: 10.2.10.91
        - nvidia-cusolver-cu11: 11.4.0.1
        - nvidia-cusparse-cu11: 11.7.4.91
        - nvidia-nccl-cu11:  2.14.3
        - nvidia-nvtx-cu11:  11.7.91
        - pip:               20.0.2
        - pkg-resources:     0.0.0
        - setuptools:        44.0.0
        - sympy:             1.12
        - torch:             2.0.1
        - triton:            2.0.0
        - typing-extensions: 4.5.0
        - wheel:             0.34.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - release:           5.4.0-137-generic
        - version:           #154-Ubuntu SMP Thu Jan 5 17:03:22 UTC 2023

</details>

cc @svekars @carljparker @mruberry @rgommers

**Comment by Verma-Rajat** [comment]
For the same data and axis, np.unique returns data in correctly sorted order:

```python
import numpy as np
x=np.array([[[[1, 0, 1, 0, 1, 1],
         [0, 1, 1, 0, 1, 1]],
        [[0, 1, 1, 0, 0, 1],
         [0, 0, 0, 1, 0, 0]]],
       [[[0, 1, 0, 1, 1, 1],
         [0, 1, 1, 0, 1, 1]],
        [[0, 0, 1, 1, 0, 1],
         [1, 1, 0, 0, 0, 0]]]])
y=np.unique(x,axis=-1)
print(y)
```

Output:
0, 0, 1, **1**, **1**, 1
0, 1, 0, **1**, **1**, 1
0, 1, 0, **0**, **1**, 1
1, 0, 0, 0, 0, 0
1, 1, 0, 1, 0, 1
0, 1, 0, 1, 1, 1
1, 0, 0, 0, 1, 1
0, 1, 1, 0, 0, 0

**Comment by nikitaved** [comment]
I totally agree, @Verma-Rajat , what numpy is doing makes more sense to me, but `torch.unique` does something totally different, and the documentation does not really help, so I am not sure whether it is a bug or a feature...

**Comment by nikitaved** [comment]
@Verma-Rajat , as a work-around for now, try using 2D inputs, there the algo is fine and it is doing what is expected:
```python
In [4]: torch.unique(x.reshape(-1, 6), dim=1)
Out[4]: 
tensor([[0, 0, 1, 1, 1, 1],
        [0, 1, 0, 1, 1, 1],
        [0, 1, 0, 0, 1, 1],
        [1, 0, 0, 0, 0, 0],
        [1, 1, 0, 1, 0, 1],
        [0, 1, 0, 1, 1, 1],
        [1, 0, 0, 0, 1, 1],
        [0, 1, 1, 0, 0, 0]])

```
